File: app/src/main/java/com/amibar/boggle/ui/mainmenu/FriendListActivity.java
package com.amibar.boggle.ui.mainmenu;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;

import com.amibar.boggle.R;
import com.amibar.boggle.data.FirebaseHandler;
import com.amibar.boggle.data.User;
import com.amibar.boggle.databinding.ActivityFriendlistBinding;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ServerValue;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Activity for managing and viewing a user's friend list.
 * Allows users to search for others by email, add friends, and invite them to game rooms.
 * Uses Firebase Realtime Database for all persistence.
 */
public class FriendListActivity extends AppCompatActivity {

    /** Tag used for logging. */
    private static final String TAG = "FriendListActivity";
    /** View binding for the activity. */
    private ActivityFriendlistBinding binding;
    /** Adapter for the friends list RecyclerView. */
    private FriendAdapter adapter;
    /** Local list of friend objects fetched from the database. */
    private final List<User> friendsList = new ArrayList<>();
    /** Singleton instance of the FirebaseHandler. */
    private FirebaseHandler firebaseHandler;

    /** Cached snapshot of all users for searching purposes. */
    private DataSnapshot usersSnapshot;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_friendlist);
        firebaseHandler = FirebaseHandler.getInstance();

        setupRecyclerView();
        setupClickListeners();
        setupSearchInput();

        loadUsers();
    }

    /**
     * Loads the global user list (for searching) and the current user's friends list.
     */
    private void loadUsers() {
        firebaseHandler.getRootRef().child("users").get().addOnSuccessListener(snapshot -> {
            usersSnapshot = snapshot;
        }).addOnFailureListener(e -> Log.e(TAG, "Failed to load users", e));
        loadFriends();
    }

    /**
     * Initializes the RecyclerView for displaying friends and its adapter.
     */
    private void setupRecyclerView() {
        adapter = new FriendAdapter(this::showInviteDialog);
        binding.friendsRecyclerView.setAdapter(adapter);
    }

    /**
     * Displays a dialog to invite a friend to a specific game room by entering a code.
     * @param friend The user object to invite.
     */
    private void showInviteDialog(User friend) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Invite " + friend.getDisplayName());
        builder.setMessage("Enter room code to invite them to play:");

        final EditText input = new EditText(this);
        input.setHint("Room Code");
        builder.setView(input);

        builder.setPositiveButton("Send", (dialog, which) -> {
            String roomCode = input.getText().toString().trim();
            if (!roomCode.isEmpty()) {
                sendInvitation(friend, roomCode);
                // After sending, transition the host to the MainActivity which will open the room
                Intent intent = new Intent(this, MainActivity.class);
                intent.putExtra("roomCode", roomCode);
                intent.putExtra("action", "host");
                startActivity(intent);
            } else {
                Toast.makeText(this, "Room code cannot be empty", Toast.LENGTH_SHORT).show();
            }
        });
        builder.setNegativeButton("Cancel", (dialog, which) -> dialog.cancel());

        builder.show();
    }

    /**
     * Sends a game invitation record to the recipient's invitations node in Firebase.
     * This will trigger an FCM notification via the InvitationService.
     * @param friend   The recipient of the invitation.
     * @param roomCode The room code the recipient should join.
     */
    private void sendInvitation(User friend, String roomCode) {
        String currentUserId = firebaseHandler.getCurrentUserId();
        User currentUser = firebaseHandler.getUserData();
        
        if (currentUserId == null || currentUser == null) {
            Toast.makeText(this, "Error: You must be logged in", Toast.LENGTH_SHORT).show();
            return;
        }

        DatabaseReference invitationsRef = firebaseHandler.getRootRef()
                .child("invitations")
                .child(friend.getUid())
                .push();

        Map<String, Object> invitation = new HashMap<>();
        invitation.put("senderId", currentUserId);
        invitation.put("senderName", currentUser.getDisplayName());
        invitation.put("message", "Join my Boggle game!");
        invitation.put("roomCode", roomCode);
        invitation.put("timestamp", ServerValue.TIMESTAMP);

        invitationsRef.setValue(invitation)
                .addOnSuccessListener(aVoid -> Toast.makeText(this, "Invitation sent to " + friend.getDisplayName(), Toast.LENGTH_SHORT).show())
                .addOnFailureListener(e -> Toast.makeText(this, "Failed to send invitation", Toast.LENGTH_SHORT).show());
    }

    /**
     * Sets up click listeners for the refresh and add friend UI elements.
     */
    private void setupClickListeners() {
        binding.refreshButton.setOnClickListener(v -> {
            loadUsers();
            loadFriends();
        });
        binding.addFriendButton.setOnClickListener(this::addFriend);
    }

    /**
     * Adds the currently searched user as a friend in the database.
     * @param view The clicked view.
     */
    private void addFriend(View view) {
        String friendId = binding.getSearchedUser().getUid();
        if (friendId.isEmpty()) {
            Toast.makeText(this, "Please enter a valid email", Toast.LENGTH_SHORT).show();
            return;
        }
        firebaseHandler.addFriend(friendId);
        binding.friendEmailInput.setText("");
        Toast.makeText(this, "Friend added!", Toast.LENGTH_SHORT).show();
        
        // Refresh local friend list
        loadFriends();
    }

    /**
     * Configures the search input field with a TextWatcher for live user filtering.
     */
    private void setupSearchInput() {
        binding.friendEmailInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (usersSnapshot == null) return;
                
                String query = s.toString().toLowerCase();
                List<User> filteredList = new ArrayList<>();
                for (DataSnapshot userSnapshot : usersSnapshot.getChildren()) {
                    User user = userSnapshot.getValue(User.class);
                    if (user != null && user.getEmail().toLowerCase().contains(query)) {
                        // Don't show current user in search results
                        if (!user.getUid().equals(firebaseHandler.getCurrentUserId())) {
                            filteredList.add(user);
                        }
                    }
                }
                filteredList.sort((u1, u2) -> u1.getDisplayName().compareToIgnoreCase(u2.getDisplayName()));
                // Update data binding for the searched user UI
                binding.setSearchedUser(filteredList.isEmpty() ? null : filteredList.get(0));
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });
    }

    /**
     * Loads the current user's friend list from Firebase.
     */
    private void loadFriends() {
        DatabaseReference userRef = firebaseHandler.getUserRef();
        if (userRef == null) return;

        DatabaseReference friendsRef = userRef.child("friends");
        friendsRef.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                friendsList.clear();
                if (!snapshot.exists()) {
                    adapter.submitList(new ArrayList<>(friendsList));
                    return;
                }
                for (DataSnapshot friendSnapshot : snapshot.getChildren()) {
                    String friendId = friendSnapshot.getKey();
                    if (friendId != null) {
                        fetchFriendData(friendId);
                    }
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                Log.e(TAG, "Failed to load friends", error.toException());
            }
        });
    }

    /**
     * Fetches details for a specific friend ID and updates the list.
     * @param friendId The UID of the friend to fetch.
     */
    private void fetchFriendData(String friendId) {
        firebaseHandler.getRootRef().child("users").child(friendId).get()
                .addOnSuccessListener(dataSnapshot -> {
                    User friend = dataSnapshot.getValue(User.class);
                    if (friend != null) {
                        // Avoid duplicates in the local list
                        boolean exists = false;
                        for (User u : friendsList) {
                            if (u.getUid().equals(friend.getUid())) {
                                exists = true;
                                break;
                            }
                        }
                        if (!exists) {
                            friendsList.add(friend);
                            adapter.submitList(new ArrayList<>(friendsList));
                        }
                    }
                });
    }
}

File: app/src/main/java/com/amibar/boggle/ui/mainmenu/FriendListActivity.kt
package com.amibar.boggle.ui.mainmenu

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.text.input.TextFieldLineLimits
import androidx.compose.foundation.text.input.TextFieldState
import androidx.compose.foundation.text.input.rememberTextFieldState
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TextField
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.amibar.boggle.R
import com.amibar.boggle.data.User
import com.amibar.boggle.ui.game.multiplayer.Player
import com.amibar.boggle.ui.theme.BoggleTheme
import kotlinx.coroutines.launch

/**
 * Activity for managing and viewing a user's friend list.
 * Allows users to search for others by email, add friends, and invite them to game rooms.
 * Uses Firebase Realtime Database for all persistence.
 */
class FriendListActivity : AppCompatActivity() {
    private val viewModel: FriendListViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            FriendListScreen(
                viewModel = viewModel,
                onInviteFriend = {

                }
            )
        }
        observeViewModel()
    }

    private fun observeViewModel() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    viewModel.events.collect { event ->
                        when (event) {
                            is FriendListEvent.ShowToast -> Toast.makeText(
                                this@FriendListActivity,
                                event.message,
                                Toast.LENGTH_SHORT
                            ).show()

                            is FriendListEvent.NavigateToHostGame -> {
                                val intent = Intent(this@FriendListActivity, MainActivity::class.java)
                                intent.putExtra("roomCode", event.roomCode)
                                intent.putExtra("action", "host")
                                startActivity(intent)
                            }
                        }
                    }
                }
            }
        }
    }

}

@Composable
fun FriendListScreen(
    modifier: Modifier = Modifier,
    viewModel: FriendListViewModel,
    onInviteFriend: (User) -> Unit = {}
) {
    val searchedUser by viewModel.searchedUser.collectAsState()
    val friends by viewModel.friends.collectAsState()
    val friendToInvite = remember { mutableStateOf<User?>(null) }

    FriendListScreenContent(
        searchedUser = searchedUser,
        friends = friends,
        searchQueryState = viewModel.searchQueryState,
        onAddFriend = viewModel::addFriend,
        onRefresh = viewModel::loadData,
        onInviteFriend = {
            friendToInvite.value = it
            onInviteFriend(it)
        },
        modifier = modifier
    )

    friendToInvite.value?.let { friend ->
        InviteDialog(
            friend = friend,
            onDismissRequest = { friendToInvite.value = null },
            onInviteSent = { user, roomCode ->
                viewModel.sendInvitation(user, roomCode)
                friendToInvite.value = null
            }
        )
    }
}

@Composable
fun FriendListScreenContent(
    friends: List<User>,
    onAddFriend: () -> Unit,
    onRefresh: () -> Unit,
    onInviteFriend: (User) -> Unit,
    modifier: Modifier = Modifier,
    searchedUser: User? = null,
    searchQueryState: TextFieldState = rememberTextFieldState(),
) {
    Column(
        modifier = modifier
            .padding(16.dp)
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            TextField(
                state = searchQueryState,
                label = { Text("Email") },
                modifier = Modifier.weight(1f)
            )
            Button(
                modifier = Modifier.padding( start = 8.dp ),
                onClick = onAddFriend
            ) {
                Text("Add")
            }
        }

        Row(
            modifier = Modifier.fillMaxWidth()
                .heightIn( max = 32.dp ),
            horizontalArrangement = Arrangement.End,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            searchedUser?.let { user ->
                Player(
                    player = user,
                    modifier = Modifier
                        .weight(1f)
                        .background(MaterialTheme.colorScheme.surfaceVariant)
                )
            }
            TextButton(
                onClick = onRefresh,
                modifier = Modifier.padding( start = 8.dp )
            ) {
                Text("Refresh")
            }
        }

        LazyColumn {
            items(friends) { friend ->
                Friend(
                    friend = friend,
                    onClickInvite = { onInviteFriend(friend) },
                    modifier = Modifier.fillMaxWidth()
                )
            }
        }
    }
}

@Composable
fun Friend(friend: User, modifier: Modifier = Modifier, onClickInvite: () -> Unit = {}) {
    Row(
        modifier = modifier,
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Player(
            player = friend,
            modifier = Modifier.weight(1f)
        )
        Button(
            onClick = onClickInvite
        ) {
            Text("invite")
        }
    }
}

/**
 * Displays a composable dialog that allows the user to invite a friend to a game.
 * Includes a text field for entering a room code and actions to send the invitation or cancel.
 *
 * @param modifier The [Modifier] to be applied to the dialog.
 * @param friend The [User] being invited.
 * @param onDismissRequest Callback invoked when the user attempts to dismiss the dialog.
 * @param onInviteSent Callback invoked with the [User] object when the "Send" button is clicked.
 */
@Composable
private fun InviteDialog(
    modifier: Modifier = Modifier,
    friend: User,
    onDismissRequest: () -> Unit = {},
    onInviteSent: (User, String) -> Unit = { _, _ -> }
) {
    val roomCodeState = rememberTextFieldState()
    androidx.compose.material3.AlertDialog(
        modifier = modifier,
        title = {
            Text(text = "Invite ${friend.displayName}?")
        },
        text = {
            TextField(
                state = roomCodeState,
                lineLimits = TextFieldLineLimits.SingleLine,
                label = {
                    Text("Room Code")
                }
            )
        },
        onDismissRequest = onDismissRequest,
        confirmButton = {
            Button(
                onClick = { onInviteSent(friend, roomCodeState.text.toString()) }
            ) {
                Text("Send")
            }
        },
        dismissButton = {
            TextButton(
                onClick = onDismissRequest
            ) {
                Text(
                    text = "Cancel"
                )
            }
        }
    )
}
